chain/chain/src/chain.rs

```diff
@@ -538,21 +535,110 @@ impl Chain {
+    // 2. Let's define clearing height = 103. It this case Fork Clearing will be executed for A
+    //    to delete blocks G and F, then Fork Clearing will be executed for B to delete block E.
+    //    Then Canonical Chain Clearing will delete blocks A and B as unlocked.
+    //    Block C is the only block of height 103 remains on the Canonical Chain (invariant).
+    //
+    pub fn clear_data(&mut self, trie: Arc<Trie>) -> Result<(), Error> {
         let mut chain_store_update = self.store.store_update();
         let head = chain_store_update.head()?;
-        let gc_head_height = chain_store_update.gc_head_height()?;
+        let tail = chain_store_update.tail()?;
         let gc_stop_height = self.runtime_adapter.get_gc_stop_height(&head.last_block_hash)?;
-        let new_gc_head_height = min(gc_head_height + HEIGHTS_TO_CLEAR + 1, gc_stop_height);
-        for height in gc_head_height..new_gc_head_height {
-            match chain_store_update.clear_old_data_on_height(height) {
-                Ok(_) => {}
-                Err(err) => {
-                    error!(target: "client", "Error clearing old data on height {:?}, {:?}", height, err);
+        if gc_stop_height > head.height {
+            return Err(ErrorKind::GCError(
+                "gc_stop_height cannot be larger than head.height".into(),
+            )
+            .into());
+        }
+        // Forks Cleaning
+        for height in tail..gc_stop_height {
+            if height == chain_store_update.get_genesis_height() {
+                continue;
+            }
+            chain_store_update.clear_forks_data(trie.clone(), height)?;
+        }
+
+        // Canonical Chain Clearing
+        for height in tail + 1..gc_stop_height {
+            let blocks_current_height: Vec<CryptoHash> =
+                match chain_store_update.get_all_block_hashes_by_height(height) {
+                    Ok(blocks_current_height) => {
+                        blocks_current_height.values().flatten().cloned().collect()
+                    }
+                    _ => continue,
+                };
+            if blocks_current_height.len() != 0 {
+                if blocks_current_height.len() > 1 {
+                    // Fork is here
+                    break;
+                }
+                let block_hash = blocks_current_height[0];
+                let prev_hash = chain_store_update.get_block_header(&block_hash)?.prev_hash;
+                if *chain_store_update.get_block_refcount(&prev_hash)? != 1 {
+                    return Err(ErrorKind::GCError(format!(
+                        "invalid refcount for {:?}, expected 1, found {:?}",
+                        block_hash,
+                        chain_store_update.get_block_refcount(&prev_hash).unwrap()
+                    ))
+                    .into());
                 }
+                chain_store_update.clear_block_data(trie.clone(), block_hash, false)?;
```

Comment:
Can we [avoid having `bool`s in "public" APIs](https://www.luu.io/posts/dont-use-booleans/)? It is hard to reason about the code.

Reply:
It doesn't look like public API.